### ISSUE TITLE
How to achieve uniform behaviour across platforms?

### ISSUE THREAD

**Issue by vog** [mxe/mxe#145]
_I noticed that lots of the issues of MXE could be solved if MXE would behave more uniformly across all those different platforms. I wonder if we could fix those issues with a better design. This GitHub issue is meant to collect ideas and opinions on the future design of MXE._

Although MXE is faily portable, there are always (and have been) various issues that appear on one platform but not on another. The means people are accidently breaking the build for others without even noticing. Some past problems were even depending on which other software packages are installed on the build system. And I fear that there are possibly lots of subtle problems that nobody noticed yet, waiting to be found the hard way.

The previous testing phases were mostly about different problems of various distros, instead of about whether the cross-compiled applications actually work. Also, the current _master_ has some hairy problems so that I wonder if we should even bother announcing another testing phase. On the other hand, the current testing branch became quite old (#136), so it becomes a less and less useful alternative to the master branch.

Of course we could approach those issues by having virtual machines (VMs) for any distro, which check that all packages build on all platforms, and which probably even automatically file bug reports in the case of discrepancies. But even then, we'd had to create portability fixes one by one for each package on each platform. But in the end, we have just one target platforms (or two, if you count Win32 and Win64 separately).

So I wonder if we could fix those issues with a better design, rather than a piling stack of portability fixes.

The following proposals come to my mind:

**1. Design MXE for a single VM**

As an extreme idea, we could design MXE for a single, specific virtual machine. At the risk of starting a distro war, let's assume we're using Debian. It is a well maintained, easily installed binary distro, and I'm familar with it ;-). So we'd provide a disk image that anyone can download and run within VirtualBox, KVM or whatever. It would contain a script that updates the Debian system (if needed), and runs the latest MXE. Maybe we'd even have a bootstrap script that builds a Debian VM from scratch such that it fulfils all MXE requirements. _Upside:_ An identical system for everyone, full reproducibility, almost no chance to introduce a bug that affects other but not yourself. _Downside:_ Only really usable on systems with hardware virtualization, otherwise too slow. Also, triggering an automatic build and moving the resulting binaries from the VM is a bit cumbersome. One would have to use SSH or SSHFS or similar.

**2. Design MXE for Chroot, LFS-like**

We could also design MXE more LFS-like (http://www.linuxfromscratch.org/). So we'd build a full new toolchain within a new Chroot environment we create, with everything compiled from source, ensuring that all native libraries/tools are in the same version, and built from the same sources, as their cross variants. We wouldn't use any pre-installed native tools except for bootstrapping. To build a package, one would chroot into that environment and use that toolchain, disconnected from the main system. _Upside:_ An identical system for everyone, full reproducibility. Also, copying from and to the chroot environment is trivial, as it is in the same file system. Automatic tasks simply call some scripts through chroot. _Downside:_ Chroot usually requires root privileges, but you could drop those provileges to normal user after entering. Also, it is possible that this chroot environment would work only on Linux systems and not on MacOSX or FreeBSD. Or, maybe some tweaks are necessary for Mac and/or BSD, but then the environment would be different on those platforms, which is what we wanted to avoid in the first place. So in the worst case, Mac and BSD users would have to use a Linux VM to build and use this Chroot environment.

**3. Use existing build systems**

We could make use of existing infrastructure such as Fedora's MinGW packages, openSUSE's MinGW packages or Debian's dpkg-cross. We'd have to decide on one of those platforms, and reshape our src/_.mk files to be valid build files for that platform. *Upside:_ Using existing infrastructure, such as standardized source package formats and existing build farms. _Downside:_ Lots of overhead that made me create MXE in the first place. :-) Also, people with a different distro would probably still have to use a VM, unless they get their final applications (EXE files) built by those build farms, too.

 **Please share your opinion, refine those ideas, and/or provide more ideas.**

**Comment by lbguilherme** [comment]
One proposal could be to use a docker container for building MXE. This is somewhat equivalent to using a VM, but without the overhead and slowness on hardware without virtualization. The only requirement here would be to run a Linux kernel version 3.10 (Jun 2013) or newer (I believe FreeBSD and some others are supported too, not sure). For systems with other operating systems or older Linuxes, a VM could be provided to simply run docker (there are many ready images for that, boot2docker for instance).

The pros here are a fully controlled and reproducible environment, with easy methods of exchanging files with the host. Things could run as simply as:

    docker run mxe i686-w64-mingw32.shared-gcc main.c -o main.exe

I don't really see a downside for this.

**Comment by vog** [comment]
I agree that Docker has become an option. It is a nice alternative to a hand-written `chroot` or `LXC` / `jails` environment. However, I'm still a bit suspicious about relying on Docker, as it is some kind of container standard with just one implementation. But I guess that's is just a matter of personal taste.

The exact technology is up to whoever will sit down and implement that thing. :-)

Would anyone volunteer to implement that thing in Docker? (or any other technology?)

**Comment by lbguilherme** [comment]
I agree that Docker is not perfect, but right now is the most mature option, the only viable one IMHO. Anyway, a implementation using it would be convertible to any other container alternative, they don't differ too much.

I'll be working on a prototype of this.

**Comment by TimothyGu** [comment]
I big +1 from me. I would prefer if the base system is Debian since that is what I'm most used to. Docker is open-source and on the rising edge, so I wouldn't worry about compatibility for now :)

**Comment by vog** [comment]
+1 for Debian/Stable (currently Jessie) from my side, too.

**Comment by lbguilherme** [comment]
@TimothyGu @vog Debian was indeed my choice too.

Here is the initial prototype: https://github.com/lbguilherme/dmxe

The idea is to provide a command line tool for two tasks: Creating images and running them,
where image has a single MXE_TARGET and a set of mxe libraries. Updating anything involves recreating the image from scratch (slower, but will always be very clean). When running the image as a container, it will mount the current working directory as a volume and open a shell directly on it. So running commands there will affects the files outside. I've also added symbolic links to all binaries on mxe/usr/bin removing the target part. So `gcc` points to `i686-w64-mingw32.static-gcc`. Note: Currently there is no way of debugging an error, all logs are simply lost.

I didn't try compiling anything non-trivial yet, I'm mostly looking at the design of it for now. What do you think?

**Comment by vog** [comment]
This looks very promising. Nice work! However, I have two minor objections to the current state, though:

1. Do we really need 3 executables here? (`dmxe`, `dmxe-build` and `dmxe-run`). I'd prefer a single executable, named `dmxe`. But maybe there are important reasons to split it?
2. I'm afraid the prefixless symlinks will to more harm than good. With that in place, the application's build system may believe it is in a native build rather than a cross build. Also, many applications want access to both, the native and the cross compiler. (e.g. to run a code generator written in C/C++, or if you have a compiler project that needs to bootstrap itself)

If those issues are fixed, I propose to include it into MXE's `tools/` directory, and to put the contents of your README.md into a new section of MXE's documentation (`index.html`). Finally, announce it at the mailing list and encourage people to try `tools/dmxe` for compiling their application(s).

**Comment by tonytheodore** [comment]
> Here is the initial prototype: https://github.com/lbguilherme/dmxe

Great start! Works as expected on OSX using the new [docker toolbox](https://docs.docker.com/installation/mac/) (which also runs on windows).

**Comment by lbguilherme** [comment]
@vog:
1. Not really, it was more for organization and it is indeed better to merge them. Done.
2. You are right here, I removed the symbolic links and also allowed more than one target on the same image.

---

I added support for pulling the repo and rebuilding just what changed, thus it doesn

**Comment by UniqueFool** [comment]
Regarding the docker comments, some of you may find this interesting: https://www.packer.io/intro

**Comment by ashkulz** [comment]
I just proposed an approach for building packages more regularly in #2021, and if that gets implemented it's trivial to create a docker image: just include the repository in `sources.list.d` -- no need to push an image to Docker Hub.

For other platforms (e.g. macOS / FreeBSD), you do need a VM-based approach. I think `vagrant` would be a good choice, like I'm doing with [wkhtmltopdf](https://github.com/wkhtmltopdf/packaging).

**Comment by Zirias** [comment]
Another possibility could be configurable build options, and for FreeBSD, maybe use jails for building packages.

I just managed to cross-compile my work-in-progress Qt application for windows, fully statically linked, using the latest MXE on FreeBSD. Now that it works, it's awesome. I had the following issues:

- building with `clang` causes several problems. MXE doesn't find GCC because FreeBSD ports of GCC only install binaries named like `gcc8`, `g++9`, etc. I had to create symlinks on my system myself, same thing for `python -> python2.7`. Some configuration options and a recommendation to install and use a GCC port on FreeBSD would help with this.
- For glib, I had to add building of the `kqueue` part. This is only enabled for darwin systems, but FreeBSD needs it to. That's where I think a configurable build option might come in handy -- easier than trying to autodetect what has to be done on any system.
- Another similar thing: I couldn't build `cmake` at first because it tries to build a bundled `curl`, which tries to link against the system-wide OpenSSL. I use libressl on my system instead and `curl` is known to be incompatible with libressl. In my case, adding a `--system-curl` in the `cmake` package solved this issue. Again something that one could make configurable.
- I had yet another problem with `cmake` because some stuff on my host system in `/usr/local/lib/cmake` broke the build. This is where I think building in a jail would have a huge benefit. Unfortunately, this would be a FreeBSD-only thing and require quite some work to get it running :(

I don't really like the idea to use a vm, as this vm would probably also be needed to **use** MXE, and one great benefit of MXE for me is that I can do my native and cross builds from the same source tree without even switching terminals. In fact, adding Linux as a cross-compilation target would be awesome ;)

**Comment by vog** [comment]
I wouldn't mind adding FreeBSD-specific changes to MXE, but would prefer to have them generically.

Example: It should use "g++9" instead of "g++" not because the system is FreeBSD, but because the system has no "g++" but "g++9".

Same for Python, etc.

Another example: `--system-curl` could be added unconditionally, as I'm pretty sure this will work on all Linux systems as well (but we should test that nevertheless, of course).

So, @Zirias, feel free to create pull requests which fix some or all of the issues you identified.
